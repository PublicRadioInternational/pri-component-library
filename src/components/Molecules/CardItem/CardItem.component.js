/**
 * @file CardItem.component.js
 * Exports a card item component.
 */

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './CardItem.css';

import Icon from '../../Atoms/Svg/Icons.component';

const cx = classNames.bind(styles);

/**
 * Renders it's children, linked to a given url.
 */
const LinkedItem = ({ url, children, className }) => {
  if (url) {
    return (
      <a className={className} href={url}>
        {children}
      </a>
    );
  }

  return children || null;
};

LinkedItem.propTypes = {
  url: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.node.isRequired
};

LinkedItem.defaultProps = {
  url: null,
  className: null
};

/**
 * Renders blurb content in a paragraph or a div, if freeform is true.
 */
const BlurbContent = ({ freeform, children, className }) => {
  if (freeform) {
    return <div className={className}>{children}</div>;
  }

  return <p className={className}>{children}</p>;
};

BlurbContent.propTypes = {
  freeform: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

BlurbContent.defaultProps = {
  className: null
};

/**
 * Component that renders a Card Item.
 */
const CardItem = ({
  url,
  title,
  imgSrc,
  imgAlt,
  blurb,
  large,
  hasAudio,
  freeform
}) => {
  const largeClasses = element =>
    cx({
      [element]: true,
      [`${element}Lg`]: large
    });

  const freeformClasses = element =>
    cx({
      [element]: freeform
    });

  return (
    <article
      className={largeClasses('cardItem')}
      typeof="sioc:Item foaf:Document"
    >
      <div className={largeClasses('titleWrap')}>
        {title && (
          <h2
            className={`${!large ? styles.title : ''} ${freeformClasses(
              'freeformTitle'
            )}`}
          >
            <LinkedItem
              url={url}
              className={`${styles.link} ${freeformClasses('freeformLink')}`}
            >
              {title}
            </LinkedItem>
          </h2>
        )}
        {title && <span property="dc:title" content={title} />}
        <span property="sioc:num_replies" content="0" datatype="xsd:integer" />
      </div>
      {imgSrc && (
        <figure className={largeClasses('image')}>
          <LinkedItem url={url}>
            <img
              typeof="foaf:Image"
              src={imgSrc}
              alt={imgAlt}
              className={largeClasses('img')}
            />
          </LinkedItem>
        </figure>
      )}
      <BlurbContent freeform={freeform} className={largeClasses('blurb')}>
        {blurb}
        {hasAudio && (
          <a className={styles.iconLink} href={url}>
            <Icon
              name="volume"
              className={styles.icon}
              isRoundIcon
              ariaLabel="Audio"
            />
          </a>
        )}
      </BlurbContent>
    </article>
  );
};

CardItem.propTypes = {
  url: PropTypes.string,
  title: PropTypes.string,
  imgSrc: PropTypes.string,
  imgAlt: PropTypes.string,
  blurb: PropTypes.node,
  large: PropTypes.bool,
  hasAudio: PropTypes.bool,
  freeform: PropTypes.bool
};

CardItem.defaultProps = {
  imgSrc: null,
  imgAlt: null,
  blurb: null,
  url: null,
  title: null,
  large: false,
  hasAudio: false,
  freeform: false
};

export default CardItem;
