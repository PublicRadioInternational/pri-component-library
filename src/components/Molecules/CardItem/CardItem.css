/* import colors... */
@value colors: "../../00_global/colors.css";
@value grayLighter, grayDark, blue, gray, orange from colors;

/* import breakpoints */
@value min as bp-min from "../../00_global/breakpoints.css";

.cardItem {
  border-bottom: 1px solid grayLighter;
  display: flex;
  flex-wrap: wrap;
  padding: 1rem;
}

.cardItemLg {
  padding: 0 0 2rem;
}

@media bp-min {
  .titleWrap {
    flex: 0 1 63%;
    margin-right: 4%;
  }
}

.titleWrapLg {
  flex: 0 1 100%;
  margin-right: 0;
  padding: 1rem 1rem 0;
}

.title {
  font-size: 1.5rem;
}

.blurb h2 a,
.link {
  color: gray;
  text-decoration: none;
}

.link:hover {
  color: orange;
}

.image {
  margin: 0;
  flex: 0 1 30%;
  margin-left: 3%;
}

.imageLg {
  flex: 0 1 100%;
  order: -1;
  margin-left: 0;
}

.img {
  display: none;
}

@media bp-min {
  .img {
    display: block;
  }
}

.imgLg {
  display: block;
  width: 100%;
}

.blurb {
  flex: 0 1 100%;
  margin-top: 1rem;
}

.blurbLg {
  padding: 0 1rem;
}

.blurb a {
  text-decoration: none;
}

.blurb a:hover {
  color: orange;
}

.blurb p {
  margin-top: 1rem;
}

.iconLink {
  color: white;
  position: relative;
}

.iconLink:hover {
  color: grayLighter;
}

.icon {
  left: 3px;
  position: absolute;
  top: 1px;
}
