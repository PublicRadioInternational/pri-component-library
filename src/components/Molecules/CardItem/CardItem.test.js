/**
 * @file Card.test.js
 * Contains tests for Card.component.js.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import CardItem from './CardItem.component';

describe('<CardItem />', () => {
  it('Matches the Card Item snapshot', () => {
    const component = renderer
      .create(<CardItem url="/" title="Test Title" />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
