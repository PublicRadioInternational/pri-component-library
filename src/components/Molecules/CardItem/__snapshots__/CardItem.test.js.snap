// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardItem /> Matches the Card Item Default snapshot 1`] = `
<article
  className="cardItem undefined"
  typeof="sioc:Item foaf:Document"
>
  <div
    className="titleWrap undefined"
  >
    <h2
      className="title"
    >
      <a
        className="link"
        href="/"
      >
        Test Title
      </a>
    </h2>
    <span
      content="Test Title"
      property="dc:title"
    />
    <span
      content="0"
      datatype="xsd:integer"
      property="sioc:num_replies"
    />
  </div>
  <figure
    className="image undefined"
  >
    <a
      href="/"
    >
      <img
        alt={null}
        className="img undefined"
        src={null}
        typeof="foaf:Image"
      />
    </a>
  </figure>
  <p
    className="blurb undefined"
  />
</article>
`;

exports[`<CardItem /> Matches the Card Item Large snapshot 1`] = `
<article
  className="cardItem cardItemLg"
  typeof="sioc:Item foaf:Document"
>
  <div
    className="titleWrap titleWrapLg"
  >
    <h2
      className="undefined"
    >
      <a
        className="link"
        href="/"
      >
        Test Title
      </a>
    </h2>
    <span
      content="Test Title"
      property="dc:title"
    />
    <span
      content="0"
      datatype="xsd:integer"
      property="sioc:num_replies"
    />
  </div>
  <figure
    className="image imageLg"
  >
    <a
      href="/"
    >
      <img
        alt={null}
        className="img imgLg"
        src={null}
        typeof="foaf:Image"
      />
    </a>
  </figure>
  <p
    className="blurb blurbLg"
  />
</article>
`;
